myia/debug/utils.py

```diff
@@ -0,0 +1,102 @@
+"""Miscellaneous utilities for debugging."""
+
+from collections import defaultdict
+from myia.debug.label import short_labeler
+from myia.anf_ir import Graph
+from myia.anf_ir_utils import \
+    is_constant, is_parameter, is_apply, succ_deeper
+from myia.graph_utils import dfs, always_include
+
+
+class _Empty:
+    """Bogus class, used internally by mixin."""
+
+    pass
+
+
+def mixin(target):
+    """Class decorator to add methods to the target class."""
+    def apply(cls):
+        methods = set(dir(cls))
+        methods.difference_update(set(dir(_Empty)))
+        for mthd in methods:
+            setattr(target, mthd, getattr(cls, mthd))
+        return target
+    return apply
+
+
+def isomorphic(g1, g2, equiv=None):
+    """Return whether g1 and g2 are structurally equivalent.
+
+    Constants are isomorphic iff they contain the same value or are isomorphic
+    graphs.
+
+    g1.return_ and g2.return_ must represent the same node under the
+    isomorphism. Parameters must match in the same order.
+    """
+    if len(g1.parameters) != len(g2.parameters):
+        return False
+
+    prev_equiv = equiv
+    equiv = dict(zip(g1.parameters, g2.parameters))
+    if prev_equiv:
+        equiv.update(prev_equiv)
+
+    def same(n1, n2):
```

Comment:
This should be non-recursive.  Also, I would probably move it to anf_ir_utils since it is something that would be generally useful.